HeterogeneousCore/AlpakaInterface/interface/AllocatorPolicy.h

```diff
@@ -0,0 +1,53 @@
+#ifndef HeterogeneousCore_AlpakaInterface_interface_AllocatorPolicy_h
+#define HeterogeneousCore_AlpakaInterface_interface_AllocatorPolicy_h
+
+#include <alpaka/alpaka.hpp>
+
+#include "HeterogeneousCore/AlpakaInterface/interface/traits.h"
+
+namespace cms::alpakatools {
+
+  // Which memory allocator to use
+  //   - Synchronous:   (device and host) cudaMalloc/hipMalloc and cudaMallocHost/hipMallocHost
+  //   - Asynchronous:  (device only)     cudaMallocAsync (requires CUDA >= 11.2)
+  //   - Caching:       (device and host) caching allocator
+  enum class AllocatorPolicy { Synchronous = 0, Asynchronous = 1, Caching = 2 };
+
+  template <typename TDev, typename = std::enable_if_t<cms::alpakatools::is_device_v<TDev>>>
+  constexpr inline AllocatorPolicy allocator_policy = AllocatorPolicy::Synchronous;
+
+#if defined ALPAKA_ACC_CPU_B_SEQ_T_SEQ_ENABLED || defined ALPAKA_ACC_CPU_B_TBB_T_SEQ_ENABLED
+  template <>
+  constexpr inline AllocatorPolicy allocator_policy<alpaka::DevCpu> =
+#if !defined ALPAKA_DISABLE_CACHING_ALLOCATOR
+      AllocatorPolicy::Caching;
+#else
```

Comment:
First of all, thanks for asking: in order to answer I started looking a bit better at the use of the memory buffers and I realised the caching allocator was not being used :-/ . That should now be fixed.

Then, here is what happens when we run the test with the GPU backend:
  - `alpaka_cuda_async::TestAlpakaProducer` calls `cms::alpakatools::make_device_buffer<std::byte[]>(const alpaka::QueueCudaRtNonBlocking&, alpaka_common::Extent)`
  - `make_device_buffer(...)` calls `CachingAllocator<alpaka::DevCudaRt, alpaka::QueueCudaRtNonBlocking>::allocate(size, queue)`
  - `alpaka_cuda_async::TestAlpakaTranscriber` calls `cms::alpakatools::make_host_buffer(const alpaka::QueueCudaRtNonBlocking&, alpaka_common::Extent)`
  - `make_host_buffer(...)` calls `CachingAllocator<alpaka::DevCpu, alpaka::QueueCudaRtNonBlocking>::allocate(size, queue)`

So, it is using the device caching allocator for the GPU memory, and the host caching allocator for the pinned host memory.

When running with the CPU serial backend:
  - `alpaka_serial_sync::TestAlpakaProducer` calls `cms::alpakatools::make_host_buffer(const alpaka::QueueCpuBlocking&, alpaka_common::Extent)`, which does __not__ call the caching allocator
  - `alpaka_serial_sync::TestAlpakaTranscriber` calls `cms::alpakatools::make_host_buffer(const alpaka::QueueCpuBlocking&, alpaka_common::Extent)`, which does __not__ call the caching allocator

So, the caching allocator is not used at all.

Reply:
So, while we _could_ implement a caching allocator also for plain host memory - even if just as an exercise or to check if it helps for anything - currently that's not the case.

Reply:
Thanks for the details. So in a sense for CPU-serial backend the `Caching` policy is just implemented as non-caching, and the same policy impacts also the pinned host memory allocations for CUDA backends, where `Caching` vs `Synchronous` makes a difference. Did I get that right?

Reply:
Indeed, it's a bit confusing, and something we could eventually simplify.

Here I'm going through the code and trying to describe the current implementation - let's see if it actually makes sense ...

### Allocator policy

The default policy is `AllocatorPolicy::Synchronous`:
https://github.com/fwyzard/cmssw/blob/40695fa1210f6ff4f9c9558232a8745b43b35d3e/HeterogeneousCore/AlpakaInterface/interface/AllocatorPolicy.h#L16-L17

But for all relevant cases, the policy is actually `AllocatorPolicy::Caching`.
  -  CPU serial and TBB:
    https://github.com/fwyzard/cmssw/blob/40695fa1210f6ff4f9c9558232a8745b43b35d3e/HeterogeneousCore/AlpakaInterface/interface/AllocatorPolicy.h#L20-L23
  - CUDA:
    https://github.com/fwyzard/cmssw/blob/40695fa1210f6ff4f9c9558232a8745b43b35d3e/HeterogeneousCore/AlpakaInterface/interface/AllocatorPolicy.h#L30-L33
  - HIP/ROCm:
    https://github.com/fwyzard/cmssw/blob/40695fa1210f6ff4f9c9558232a8745b43b35d3e/HeterogeneousCore/AlpakaInterface/interface/AllocatorPolicy.h#L42-L45

### `make_host_buffer` / `make_device_buffer`

The calls to `make_host_buffer` and `make_device_buffer` that _do not_ pass a `Queue` are always non-caching:
  - https://github.com/fwyzard/cmssw/blob/40695fa1210f6ff4f9c9558232a8745b43b35d3e/HeterogeneousCore/AlpakaInterface/interface/memory.h#L61-L78
  - https://github.com/fwyzard/cmssw/blob/40695fa1210f6ff4f9c9558232a8745b43b35d3e/HeterogeneousCore/AlpakaInterface/interface/memory.h#L150-L170

In the host case, these calls are not associated to any accelerator, and return default, non-pinned, system  memory (_i.e._ `new` with alignment).
In the device case, these calls are not associated to any queue, and return global memory (_i.e._ `cudaMalloc`).

The policy is used by the calls to `make_host_buffer` and `make_device_buffer` that _do_ take a `Queue`:
  - https://github.com/fwyzard/cmssw/blob/40695fa1210f6ff4f9c9558232a8745b43b35d3e/HeterogeneousCore/AlpakaInterface/interface/memory.h#L80-L113
  - https://github.com/fwyzard/cmssw/blob/40695fa1210f6ff4f9c9558232a8745b43b35d3e/HeterogeneousCore/AlpakaInterface/interface/memory.h#L172-L216

Host buffers use either `allocCachedBuf(host, device_queue)` to go through the caching allocator, or `allocMappedBuf` to get non-cached, pinned host memory (_i.e._ `cudaMallocHost`).
Device buffers use `allocCachedBuf(device, device_queue)` and go through the caching allocator, or `allocAsyncBuf` to get non-cached, stream-ordered device memory (_i.e._ `cudaMallocAsync`), or fall back to non-cached device memory (_i.e._ `cudaMalloc`).

### `allocCachedBuf`

`allocCachedBuf(TDev, TQueue, ...)` is the alpaka-like interface to the caching allocators.

The default implementation results in a `static_assert` failure: https://github.com/fwyzard/cmssw/blob/40695fa1210f6ff4f9c9558232a8745b43b35d3e/HeterogeneousCore/AlpakaInterface/interface/CachedBufAlloc.h#L14-L24 .

In absence of other specialisations, the implementation for the CPU device falls back to stream-ordered, non-cached, host only memory: https://github.com/fwyzard/cmssw/blob/40695fa1210f6ff4f9c9558232a8745b43b35d3e/HeterogeneousCore/AlpakaInterface/interface/CachedBufAlloc.h#L29-L33 .

When CUDA is enabled, two additional specialisations are available:
  - host memory associated to a CUDA queue
  - CUDA device memory

`allocCachedBuf(DevCpu, QueueCudaRtNonBlocking, ...)` uses `getHostCachingAllocator<alpaka::QueueCudaRtNonBlocking>()` to access the host caching allocator and provide cached, queue-ordered, pinned host memory: https://github.com/fwyzard/cmssw/blob/40695fa1210f6ff4f9c9558232a8745b43b35d3e/HeterogeneousCore/AlpakaInterface/interface/CachedBufAlloc.h#L42-L60

`allocCachedBuf(DevCudaRt, QueueCudaRtNonBlocking, ...)` uses `getDeviceCachingAllocator<alpaka::DevCudaRt, TQueue>(dev)` to access the corresponding device caching allocator and provide cached, queue-ordered, global device memory: https://github.com/fwyzard/cmssw/blob/40695fa1210f6ff4f9c9558232a8745b43b35d3e/HeterogeneousCore/AlpakaInterface/interface/CachedBufAlloc.h#L65-L86

When HIP/ROCm is enabled, there are equivalent specialisation for ROCm pinned host memory and global device memory.

---

OK, so the overall structure is quite complex, but it seems consistent.
There could be some improvements:
  - the versions of `make_host_buffer` that take a `queue` could add the intermediate case of falling back to `allocAsyncBuf` to provide non-cached, queue-ordered, non-pinned host memory;
  - the various specialisations could support both blocking and non-blocking queues;
  - anything else ?

Back to the question at hand

> So in a sense for CPU-serial backend the `Caching` policy is just implemented as non-caching, and the same policy impacts also the pinned host memory allocations for CUDA backends, where `Caching` vs `Synchronous` makes a difference. Did I get that right?

If I understood correctly the question, yes, you got it right:
  - there is a single overall policy (synchronous, asynchronous, or caching);
  - allocations for the CPU backend can be synchronous (without a queue) or asynchronous (with a queue, where the "caching" policy is implemented by non-caching asynchronous allocations);
  - allocations for the CUDA and HIP/ROCm backends are asynchronous and cached, both for pinned host memory and global device memory.

Reply:
Thanks. Now I have a new follow-up question :)

> In absence of other specialisations, the implementation for the CPU device falls back to stream-ordered, non-cached, host only memory:
...
> allocations for the CPU backend can be synchronous (without a queue) or asynchronous (with a queue, where the "caching" policy is implemented by non-caching asynchronous allocations);

Does the asynchronous/stream-ordered allocation of host-only memory combined with a non-blocking queue mean that something in Alpaka itself keeps the memory alive after the Alpaka buffer has been destructed until the operations in the queue at the point of destruction have been completed? (or something along that)

Reply:
If I implemented it correctly, yes: https://github.com/alpaka-group/alpaka/blob/30d205f46d7f9235dd49b4cccd4d2daaa25e0f04/include/alpaka/mem/buf/BufCpu.hpp#L249-L287

Instead of calling `free` immediately when the buffer is destroyed, the async buffer uses a deleter that enqueues the call to `free` into the queue itself.